d3a_api_client/redis.py

```diff
@@ -0,0 +1,183 @@
+        self.client_id = client_id
+        self.is_active = False
+        self._blocking_command_responses = {}
+        self._subscribe_to_response_channels()
+        if autoregister:
+            self.register(is_blocking=False)
+
+    def _subscribe_to_response_channels(self):
+        channel_subs = {
+            f"{self._command_topics[k]}/response": self._generate_command_response_callback(k)
+            for k in Commands
+        }
+        channel_subs[f'{self.market_id}/register_participant/response'] = self._on_register
+        channel_subs[f'{self.market_id}/unregister_participant/response'] = self._on_unregister
+        channel_subs[f'{self._channel_prefix}/market_cycle'] = self._on_market_cycle
+        self.pubsub.subscribe(**channel_subs)
+        self.pubsub.run_in_thread(daemon=True)
+
+    def register(self, is_blocking=False):
+        if self.is_active:
+            raise RedisAPIException(f'API is already registered to the market.')
+
+        self.redis_db.publish(f'{self.market_id}/register_participant',
+                              json.dumps({"name": self.client_id}))
+        if is_blocking:
+            try:
+                wait_until_timeout_blocking(lambda: self.is_active, timeout=30)
+            except AssertionError:
+                raise RedisAPIException(
+                    f'API registration process timed out. Server will continue processing your '
+                    f'request on the background and will notify you as soon as the registration '
+                    f'has been completed.')
+
+    def unregister(self, is_blocking=False):
+        if not self.is_active:
+            raise RedisAPIException(f'API is already unregistered from the market.')
+
+        self.redis_db.publish(f'{self.market_id}/unregister_participant',
+                              json.dumps({"name": self.client_id}))
+        if is_blocking:
+            try:
+                wait_until_timeout_blocking(lambda: not self.is_active, timeout=30)
+            except AssertionError:
+                raise RedisAPIException(
+                    f'API unregister process timed out. Server will continue processing your '
+                    f'request on the background and will notify you as soon as the unregistration '
+                    f'has been completed.')
+
+    @property
+    def _channel_prefix(self):
+        return f"{self.market_id}/{self.client_id}"
+
+    @property
+    def _command_topics(self):
+        return {
+            Commands.OFFER: f'{self._channel_prefix}/offer',
+            Commands.BID: f'{self._channel_prefix}/bid',
+            Commands.DELETE_OFFER: f'{self._channel_prefix}/delete_offer',
+            Commands.DELETE_BID: f'{self._channel_prefix}/delete_bid',
+            Commands.LIST_OFFERS: f'{self._channel_prefix}/offers',
+            Commands.LIST_BIDS: f'{self._channel_prefix}/bids'
+        }
+
+    def wait_and_consume_command_response(self, command_type):
```

Comment:
This could also be a private function to show the user the difference between "usable functions" and private ones. Do you agree?

Reply:
Indeed, converting it to a private method. Thanks!